site/components/src/main/java/uk/nhs/digital/common/components/catalogue/filters/FiltersAndLinks.java

```diff
@@ -0,0 +1,117 @@
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FiltersAndLinks {
+    public Set<String> filters = new HashSet<>();
+    public List<CatalogueLink> links;
+    public List<String> selectedFilterKeys;
+
+    public FiltersAndLinks(final List<String> userSelectedFilterKeys, final List<CatalogueLink> links, final Filters rawFilters) {
+        List<Set<String>> orderedFiltersKeysByCategory = filterKeysSortedByCategory(rawFilters, links, userSelectedFilterKeys);
+        this.links = linksWithAnyUserSelectedFilterKeys(links, userSelectedFilterKeys).collect(toList());
+        applyFiltersToLinks(userSelectedFilterKeys, orderedFiltersKeysByCategory);
+        applyFiltersToNavKeys(userSelectedFilterKeys, links, orderedFiltersKeysByCategory);
+        updateUserSelectedFilters(userSelectedFilterKeys);
+    }
+
+    private void applyFiltersToLinks(final List<String> userSelectedFilterKeys, List<Set<String>> orderedFilterKeysByCategory) {
+        if (!userSelectedFilterKeys.isEmpty()) {
+            orderedFilterKeysByCategory.forEach(category -> {
+                List<String> userSelectedFiltersForCategory = userSelectedFilterKeys.stream().filter(category::contains).collect(toList());
+                this.links = linksWithAnyUserSelectedFilterKeys(this.links, userSelectedFiltersForCategory).collect(toList());
+            });
+        }
+    }
+
+    private void applyFiltersToNavKeys(final List<String> userSelectedFilterKeys, final List<CatalogueLink> links, List<Set<String>> orderedFilterKeysByCategory) {
+        if (!userSelectedFilterKeys.isEmpty()) {
+            AtomicReference<List<CatalogueLink>> filteredLinks = new AtomicReference<>(links);
+            orderedFilterKeysByCategory.forEach(category -> {
+                Set<String> filtersForCategoryFromLinks = filtersForCategoryFromLinks(filteredLinks.get(), category);
+                List<String> userSelectedFiltersForCategory = filtersForCategory(userSelectedFilterKeys, category);
+                List<CatalogueLink> filteredLinksForCategory = linksWithAnyUserSelectedFilterKeys(filteredLinks.get(), userSelectedFiltersForCategory).collect(toList());
+                filteredLinks.set(filteredLinksForCategory);
+                filtersForCategoryFromLinks.addAll(filtersNotInCollection(filteredLinksForCategory, filtersForCategoryFromLinks));
+                addToFilters(filtersForCategoryFromLinks);
+                Set<String> allFilteredLinksKeys = filteredLinks.get().stream().flatMap(link -> link.allTaxonomyKeysOfReferencedDoc().stream()).collect(Collectors.toSet());
+                removeFromFilters(filters.parallelStream().filter(key -> !collectionsContainKey(category, allFilteredLinksKeys, key)).collect(Collectors.toSet()));
+            });
+        } else {
+            addToFilters(allKeysFromLinks(links));
+        }
+    }
```

Comment:
As this is such a big chunk of code, please can you add some comments?

Reply:
yeah sure, it is quite a wall of code to be fair